### ISSUE TITLE
Builds fail without `docker buildx`, but dependency isn't validated

### ISSUE THREAD

**Issue by ehashman** [kubernetes/kubernetes#102822]
#### What happened:

When I ran `make update`, it failed with the following error:

```
ehashman@fedora:~/src/k8s$ make update 
Running in silent mode, run with SILENT=false if you want to see script logs.
Running in short-circuit mode; run with FORCE_ALL=true to force all scripts to run.
Running update-generated-protobuf
+++ Docker build command failed for kube-build:build-09f02a7f40-5-v1.16.5-1

unknown flag: --load
See 'docker --help'.
...
```

#### What you expected to happen:

I expected the build to successfully complete, or to be warned that I am missing a required build dependency.

#### How to reproduce it (as minimally and precisely as possible):

If you don't have docker buildx installed, you will hit this error. However, that error is masked when you pass the full command line:

```
ehashman@fedora:~/src/k8s$ docker buildx build
docker: 'buildx' is not a docker command.
See 'docker --help'
ehashman@fedora:~/src/k8s$ docker buildx build --load
unknown flag: --load
See 'docker --help'.
```

#### Anything else we need to know?:

Appears to have been introduced in https://github.com/kubernetes/kubernetes/commit/086cdfb3601536f24b127d6e1a16fdd96f468126

#### Environment:
- Kubernetes version (use `kubectl version`): HEAD (1.22.x)
- Cloud provider or hardware configuration: N/A, local build
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by BenTheElder** [comment]
> ehashman@fedora:~/src/k8s$ docker buildx build

That's not the same as in the script, which should be setting `DOCKER_CLI_EXPERIMENTAL=enabled` which ungates this command.

Stable docker versions should have buildx, at least behind this env. unfortunately the project has never validated a minimum docker version for building.

**Comment by BenTheElder** [comment]
buildx as a build requirement was actually introduced at least as back as 1.21 as well, for some of the core images. we've just standardized on it more since then (including the linked commit), so if we do add some validation we should probably backport it.

I think docker buildx has been available since 19.03?, but it appears your install is too old to have the `--load` flag in buildx.

we could also try switching to `--output=type=docker` https://docs.docker.com/engine/reference/commandline/buildx_build/#load, I'm not sure which version introduced `--load`.

**Comment by ehashman** [comment]
No difference with the flag. But I am not running upstream Docker, I'm running Docker out of the Debian repos.

```
ehashman@fedora:~/src/k8s$ docker --version
Docker version 20.10.5+dfsg1, build 55c4c88
ehashman@fedora:~/src/k8s$ DOCKER_CLI_EXPERIMENTAL=enabled docker buildx
docker: 'buildx' is not a docker command.
See 'docker --help'
```

**Comment by BenTheElder** [comment]
discussed on slack: seems debian packages don't have the buildx plugin https://github.com/docker/buildx/#binary-release, while docker-ce / docker desktop do.

we should at least be checking for the command and throwing a clear error like we do with e.g. tar.

**Comment by BenTheElder** [comment]
https://github.com/docker/docker-ce-packaging/pull/415 / https://github.com/docker/for-linux/issues/1066  👀 
https://github.com/docker-library/docker/issues/156#issuecomment-489245333

**Comment by BenTheElder** [comment]
also https://github.com/docker/buildx/pull/384

**Comment by ehashman** [comment]
Note: reverting https://github.com/kubernetes/kubernetes/commit/086cdfb3601536f24b127d6e1a16fdd96f468126 fixes the issue, so I have been building locally with that as a carry patch as a workaround.

**Comment by BenTheElder** [comment]
cc @vinayakankugoyal 🙈 apparently there are docker packages that don't ship this, which maybe changes the equations a bit, see above.

**Comment by ehashman** [comment]
tl;dr from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=989917#10

`buildx` isn't actually part of moby, but Docker Inc. ships it as part of their Docker packages in their PPA. It's a separate tool with a separate code repo.

**Comment by BenTheElder** [comment]
would it be eligible for another package?

otherwise we probably need to provide support for installing it at least to some extent, e.g. if you are on mac with ancient bash we point to installing with homebrew in the failure message that detects this (and we should also be detecting that the command is missing).

**Comment by ehashman** [comment]
You can build it from source and install it as a docker plugin by following the instructions at https://github.com/docker/buildx#with-docker-1809

I doubt it'll be part of any of the FOSS distro repos soon (although I'm happy to be proven wrong on that).

**Comment by dims** [comment]
@ehashman we don't have the yard stick that something we use should be part of FOSS distro repos. Just that it needs to be easily installable.

**Comment by dims** [comment]
Seems like you can also use the release artifacts from github as well, you don't need their ppa and you won't need to build it either.

```
$ export DOCKER_BUILDKIT=1
$ mkdir -p ~/.docker/cli-plugins
$ cd ~/.docker/cli-plugins
$ wget -O docker-buildx https://github.com/docker/buildx/releases/download/v0.5.1/buildx-v0.5.1.linux-amd64
```

**Comment by maniator** [comment]
Here is an image I created to solve not having buildx in the official image https://hub.docker.com/r/maniator/dind-buildx

**Comment by maniator** [comment]
https://hub.docker.com/r/maniator/dind-buildx has been updated to use the official binary buildx image

**Comment by manugupt1** [comment]
AL2 also does not come with buildx built in and I landed in with the same issue.
